crates/red_knot_python_semantic/src/types.rs

```diff
@@ -1401,6 +1406,112 @@ impl<'db> Type<'db> {
         }
     }
 
+    /// Returns true if this type and `other` are "exactly the same".
+    ///
+    /// Two types are "exactly the same" when:
+    ///
+    /// * Both are either `Any` or `Unknown`, or
+    /// * If they are not tuples, generics, unions or intersections:
+    ///     * They share the same internal representation.
+    /// * If they are tuples:
+    ///     * They have the same length, and
+    ///     * Each element pair are "exactly the same".
+    /// * If they are generics:
+    ///     * The "origins" are "exactly the same".
+    ///     * The argument types are "exactly the same" when compared as tuples.
+    /// * If they are unions:
+    ///     * They have the same length, and
+    ///     * For each element of the first type,
+    ///       there exist one element of the second
+    ///       such that they are "exactly the same";
+    ///       each element must not be paired more than once.
+    /// * If they are intersections:
+    ///     * The positive parts are "exactly the same" when compared as unions.
+    ///     * The negative parts are "exactly the same" when compared as unions.
+    ///
+    /// Note: `Todo != Todo`.
+    ///
+    /// This powers the `assert_type()` directive.
+    pub(crate) fn is_equals_to(self, db: &'db dyn Db, other: Type<'db>) -> bool {
+        let equal = |(first, second): (&Type<'db>, &Type<'db>)| first.is_equals_to(db, *second);
+
+        match (self, other) {
+            (Type::Todo(_), Type::Todo(_)) => false,
```

Comment:
Todo is supposed to behave like `Any` or `Unknown` in the typing system, so I think our default approach should be that all three are gradually-equivalent to each other. Is there a concrete reason you felt it necessary to make `Todo != Todo`?

Reply:
I understand the reasoning, but I think it is OK for that `assert_type` to pass, until we actually are able to properly differentiate those types. It is more important that we consistently treat `Todo` in the type system as the dynamic type, just with more provenance information.